Update symbol database such that the override keyword implies that the function is also virtual

shaneasd opened the pull request:
I raised this on the  IRC but haven't seen the response so apologies if this contradicts something said there.

See for example the following code
```
struct A
{
    virtual char x() { return 'A'; }
};

struct B : A
{
    char x() override { return 'B'; }
    B() { std::cout << x(); }
};

struct B2 : A
{
    virtual char x() override { return 'X'; }
    B2() { std::cout << x(); }
};
```

cppcheck 1.87 will correctly warn that "Virtual function 'x' is called from constructor 'B2()' at line 21. Dynamic binding is not used. [virtualCallInConstructor]" however it will not detect the same for struct B which is structurally identical but lacks the virtual keyword.

Note that B::x is already virtual by virtue of A::x being virtual and matching the method signature. The virtual keyword isn't actually doing anything in this case. It may not be obvious (I haven't checked) to cppcheck that A::x is virtual when analyzing B::x. Note that even if the code is or could be made to be smart enough to recognise this relationship, sometimes the header containing A may not be available to cppcheck at all.

It should at least recognize that B::x is virtual because it has the override keyword applied to it. Either the method is virtual or the code will not compile (which I presume would qualify as "garbage" in adhering to the cppcheck philosophy document).

This change adds a test case for the case where the override keyword is supplied and the virtual keyword is not and then modifies the implementation so this test passes by treating all methods with the override keyword as virtual.

A complete solution would require checking for the override keyword and for corresponding virtual methods in ancestors. I have done only the former here. Note however that if the override keyword is not specified and cppcheck knows that the method is overriding a parent method then there is a separate warning for not specifying the override keyword so fixing this warning would just be a time saver rather than improving coverage.

My motivation for this change is that http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c128-virtual-functions-should-specify-exactly-one-of-virtual-override-or-final recommends specifying exactly one of virtual and override which would mean hiding warnings like this. I'd like to be able to follow the cpp core guidelines but not if it means reducing cppcheck's effectiveness.

amai2012 commented:
https://trac.cppcheck.net/ticket/9179 has been created for this issue.

danmar commented:
I think it would be good to be able to differentiate if the method is implicitly or explicitly virtual. So we can for instance recommend to remove a redundant "virtual" keyword when "override" is also used. We could add a "hasVirtualKeyword" perhaps. Well.. this is not something we have to add right now it can be added later also if some check will need that info.

But as you suggested on IRC, I also think it makes sense to warn here:

    class Base {
    public:
      virtual void f() {}
    };

    class Derived : public Base {
    public:
      Derived() {
        f(); // <- warn
      }

      void f();
    };

shaneasd commented:
I think we might be able to use isImplicitlyVirtual for that. I note that isImplicitlyVirtual will return false if it fails to find all the base classes so if we check for isImplicitlyVirtual or hasOverrideSpecifier() that should be comprehensive. I'll investigate.

shaneasd commented:
I've acted on that feedback. Turns out isImplicitlyVirtual does the job nicely so this checker now calls that rather than isVirtual to determine if the method is virtual. I've modified the method to include checking the override specifier and removed the corresponding change from isVirtual so now isVirtual is just whether the virtual specifier is explicitly specified. I've updated isVirtual to hasVirtualSpecifier (matching override and final) to reflect this clearer meaning.

I have a suspicion that just about everywhere that hasVirtualSpecifier is currently being called should be calling isImplicitlyVirtual instead. However, if so, this is an existing issue exposed by the name (but not behavior) change rather than a new issue. I also think isImplictlyVirtual is misleadingly named as one of its criteria is whether the virtual keyword is provided which to me is explicit rather than implicit. I opted not to change it, however, as the obvious name would be isVirtual but making that change would make for a confusing pull request and potentially horrendous merge conflicts. Perhaps it could be renamed in 12 months or something when the possibility of merge conflicts is reduced.

This passes all the tests when I run them locally but I have a suspicion which I haven't investigated that I may have introduced an O(n^2) cost wherein we traverse the class hierarchy calling (now) isImplictlyVirtual which itself traverses the class hierarchy. I'll investigate this further but are there any performance tests which I should expect to detect this?

shaneasd commented on lib/checkclass.h:
```diff
@@ -317,7 +317,7 @@ class CPPCHECKLIB CheckClass : public Check {
 
     /**
-     * @brief gives a list of tokens where pure virtual functions are called directly or indirectly
+     * @brief gives a list of tokens where virtual functions are called directly or indirectly
```
as far as I could see the functions returned from this need not be pure virtual

shaneasd commented on lib/symboldatabase.cpp:
```diff
@@ -3144,14 +3144,16 @@ void Function::addArguments(const SymbolDatabase *symbolDatabase, const Scope *s
+    if (foundAllBaseClasses) //If we've seen all the base classes and none of the above were true then it must not be virtual
         return false;
-    return defaultVal;
+    return defaultVal; //If we can't see all the bases classes then we can't say conclusively
```
on the fence about this added documentation. It took me a while to figure out the ```if (foundAllBaseClasses) return false;``` so I thought a comment would be useful there. The rest are less necessary.

shaneasd commented:
@danmar is there any further action I need to be taking to get this merged in? I believe I've addressed your feedback.
